aten/src/ATen/native/RNN.cpp

```diff
@@ -0,0 +1,700 @@
+// transform them to fuctions of signature
+//
+// (io_type input, hidden_type hidden, param_type params) -> (io_type, hidden_type)
+//
+// which can apply the cell over a sequence of inputs, and produce both a new set
+// of hidden states, as well as a concatenated output of each step.
+
+template<typename output_type, typename hidden_type>
+struct LayerOutput {
+  output_type outputs;
+  hidden_type final_hidden;
+};
+
+template<typename io_type, typename hidden_type, typename param_type>
+struct Layer {
+  using output_type = LayerOutput<io_type, hidden_type>;
+  virtual output_type operator()(const io_type& input, const hidden_type& input_hidden, const param_type& params) const = 0;
+};
+
+template<typename hidden_type>
+struct FullLayer : Layer<Tensor, hidden_type, CellParams> {
+  using output_type = typename Layer<Tensor, hidden_type, CellParams>::output_type;
+  using unstacked_output_type = LayerOutput<std::vector<Tensor>, hidden_type>;
+
+  FullLayer(Cell<hidden_type>& cell)
+    : cell_(cell) {};
+
+  unstacked_output_type operator()(std::vector<Tensor> step_inputs, const hidden_type& input_hidden, const CellParams& params) const {
+    std::vector<Tensor> step_outputs;
+    auto hidden = input_hidden;
+    for (size_t i = 0; i < step_inputs.size(); i++) {
+      hidden = cell_(step_inputs[i], hidden, params);
+      step_outputs.push_back(hidden_as_output(hidden));
+    }
+    return {step_outputs, hidden};
+  }
+
+  output_type operator()(const Tensor& inputs, const hidden_type& input_hidden, const CellParams& params) const override {
+    auto unstacked_output = (*this)(inputs.unbind(0), input_hidden, params);
+    return {at::stack(unstacked_output.outputs, 0), unstacked_output.final_hidden};
+  }
+
+  Cell<hidden_type>& cell_;
+};
+
+template<typename dir_hidden_type>
+struct FullBidirectionalLayer : Layer<Tensor, pair_of<dir_hidden_type>, pair_of<CellParams>> {
+  using hidden_type = pair_of<dir_hidden_type>;
+  using param_type = pair_of<CellParams>;
+  using output_type = typename Layer<Tensor, hidden_type, param_type>::output_type;
+
+  FullBidirectionalLayer(Cell<dir_hidden_type>& cell)
+    : layer_(cell) {};
+
+  output_type operator()(const Tensor& input, const hidden_type& input_hidden, const param_type& params) const override {
+    auto step_inputs = input.unbind(0);
+    auto fw_result = layer_(step_inputs, input_hidden.first, params.first);
+    auto fw_output = at::stack(fw_result.outputs, 0);
+
+    auto rev_step_inputs = reverse(std::move(step_inputs));
+    auto rev_result = layer_(rev_step_inputs, input_hidden.second, params.second);
+    std::reverse(rev_result.outputs.begin(), rev_result.outputs.end());
+    auto rev_output = at::stack(rev_result.outputs, 0);
+
+    return {at::cat({fw_output, rev_output}, fw_output.dim() - 1),
+            std::make_pair(fw_result.final_hidden, rev_result.final_hidden)};
+  }
+
+  std::vector<Tensor> reverse(std::vector<Tensor>&& x) const {
```

Comment:
nit: I wouldn't call a destructive reversing function simply `reverse`

Reply:
It explicitly takes an rvalue, so the type system should prevent us from destroying non-temporary values.